File: internal/service/settlement.go
package service

import (
	"container/heap"

	"danieleambrosino.it/debt-minimizer/internal/types"
)

type MaxHeap []types.Balance

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].Amount > h[j].Amount }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x any)        { *h = append(*h, x.(types.Balance)) }
func (h *MaxHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func ComputeMinimalSettlementSet(balances []types.Balance) []types.Settlement {
	debtors := &MaxHeap{}
	creditors := &MaxHeap{}
	for _, b := range balances {
		if b.Amount < 0 {
			b.Amount = -b.Amount
			heap.Push(debtors, b)
		} else if b.Amount > 0 {
			heap.Push(creditors, b)
		}
	}

	settlements := make([]types.Settlement, 0)

	for debtors.Len() > 0 && creditors.Len() > 0 {
		d := heap.Pop(debtors).(types.Balance)
		c := heap.Pop(creditors).(types.Balance)

		amount := min(d.Amount, c.Amount)
		settlements = append(settlements, types.Settlement{
			From:   d.Person,
			To:     c.Person,
			Amount: types.UnsignedCents(amount),
		})

		d.Amount -= amount
		c.Amount -= amount

		if d.Amount > 0 {
			heap.Push(debtors, d)
		}
		if c.Amount > 0 {
			heap.Push(creditors, c)
		}
	}
	return settlements
}

File: webapp/src/lib/settlement.ts
import type { Balance, Settlement } from "./types";

class MaxHeap {
  private readonly data: Balance[] = [];

  get length(): number {
    return this.data.length;
  }

  push(val: Balance): void {
    this.data.push(val);
    this._siftUp(this.data.length - 1);
  }

  pop(): Balance {
    const top = this.data[0];
    const last = this.data.pop() as Balance;
    if (this.data.length > 0) {
      this.data[0] = last;
      this._siftDown(0);
    }
    return top;
  }

  private _siftUp(idx: number): void {
    let i = idx;
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.data[parent].amount >= this.data[i].amount) {
        break;
      }
      [this.data[parent], this.data[i]] = [this.data[i], this.data[parent]];
      i = parent;
    }
  }

  private _siftDown(idx: number): void {
    const n = this.data.length;
    let i = idx;
    while (true) {
      let largest = i;
      const left = i * 2 + 1;
      const right = left + 1;
      if (left < n && this.data[left].amount > this.data[largest].amount) {
        largest = left;
      }
      if (right < n && this.data[right].amount > this.data[largest].amount) {
        largest = right;
      }
      if (largest === i) {
        break;
      }
      [this.data[i], this.data[largest]] = [this.data[largest], this.data[i]];
      i = largest;
    }
  }
}

export function computeMinimalSettlementSet(balances: Balance[]): Settlement[] {
  const debtors = new MaxHeap();
  const creditors = new MaxHeap();
  for (const b of balances) {
    if (b.amount < 0) {
      debtors.push({ amount: -b.amount, person: b.person });
    } else if (b.amount > 0) {
      creditors.push({ amount: b.amount, person: b.person });
    }
  }

  const settlements: Settlement[] = [];

  while (debtors.length > 0 && creditors.length > 0) {
    const d = debtors.pop();
    const c = creditors.pop();

    const amount = Math.min(d.amount, c.amount);
    settlements.push({ amount, from: d.person, to: c.person });

    d.amount -= amount;
    c.amount -= amount;

    if (d.amount > 0) {
      debtors.push(d);
    }
    if (c.amount > 0) {
      creditors.push(c);
    }
  }

  return settlements;
}
